cpp/wallet_server.cc

```diff
@@ -0,0 +1,344 @@
+/*
+ *
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <unistd.h>
+#include <iostream>
+#include <memory>
+#include <string>
+
+#include <grpcpp/ext/proto_server_reflection_plugin.h>
+#include <grpcpp/grpcpp.h>
+#include <grpcpp/health_check_service_interface.h>
+
+#include "examples/protos/account.grpc.pb.h"
+#include "examples/protos/stats.grpc.pb.h"
+#include "examples/protos/wallet.grpc.pb.h"
+
+using account::Account;
+using account::GetUserInfoRequest;
+using account::GetUserInfoResponse;
+using account::MembershipType;
+using grpc::Channel;
+using grpc::ChannelArguments;
+using grpc::ClientContext;
+using grpc::ClientReader;
+using grpc::Server;
+using grpc::ServerBuilder;
+using grpc::ServerContext;
+using grpc::ServerWriter;
+using grpc::Status;
+using grpc::StatusCode;
+using stats::PriceRequest;
+using stats::PriceResponse;
+using stats::Stats;
+using wallet::BalanceRequest;
+using wallet::BalanceResponse;
+using wallet::Wallet;
+
+class WalletServiceImpl final : public Wallet::Service {
+ public:
+  void SetStatsClientStub(std::unique_ptr<Stats::Stub> stub) {
+    stats_stub_ = std::move(stub);
+  }
+
+  void SetAccountClientStub(std::unique_ptr<Account::Stub> stub) {
+    account_stub_ = std::move(stub);
+  }
+
+  void SetHostName(const std::string& hostname) { hostname_ = hostname; }
+
+  void SetV1Behavior(const bool v1_behavior) { v1_behavior_ = v1_behavior; }
+
+ private:
+  bool ObtainAndValidateUserAndMembership(ServerContext* server_context) {
```

Comment:
Part of this function can be reused with the other `ObtainAndValidateUserAndMembership`.
The difference I think is, the wallet server would need more information from this function, but the stats server just need to know premium or not.

Reply:
this suggestion is great; however because these are examples,I like the idea that all functions are contained in one file and there is no need to go to a util file or something like that.  people can just look or copy this one server file and it's a complete server.  what do you think?